packages/preferences/src/browser/abstract-resource-preference-provider.ts

```diff
@@ -59,6 +61,10 @@ export abstract class AbstractResourcePreferenceProvider extends PreferenceProvi
             this.toDisposeOnWorkspaceLocationChanged.pushAll([onDidResourceChanged, (await this.resource)]);
             this.toDispose.push(onDidResourceChanged);
         }
+
+        this.schemaProvider.onDidPreferenceSchemaChanged(() => {
+            this.handleNotValidPreferences();
```

Comment:
please collect a listener to dispose

Should not it be `this.readPreferences()`? then you can get rid of `notValidPreferences` property, `handleNotValidPreferences` method and have one place which reconcile preferences, i.e. `readPreferences` instead of have the logic in two places

Reply:
fixed